File: 1371-minimum-remove-to-make-valid-parentheses/minimum-remove-to-make-valid-parentheses.php
class Solution {

    /**
     * @param String $s
     * @return String
     */
    function minRemoveToMakeValid($s) {
        $open = 0;
        $ans = "";
        $n = strlen($s);
        for($i=0; $i<$n; ++$i){
            if($open==0 and $s[$i]==')')
                continue;
            if($s[$i] == '(')
                $open++;
            if($s[$i]==')')
                $open--;
            $ans .= $s[$i];
        }
        $n = strlen($ans);
        $ans2 = "";
        $close = 0;
        for($i=$n-1; $i>=0; --$i){
            if($close==0 and $ans[$i]=='(')
                continue;
            if($ans[$i] == ')')
                $close++;
            if($ans[$i]=='(')
                $close--;
            $ans2 .= $ans[$i];
        }
        return strrev($ans2);
    }
}

File: 1371-minimum-remove-to-make-valid-parentheses/minimum-remove-to-make-valid-parentheses.java
class Solution {
    public String minRemoveToMakeValid(String s) {
        int open = 0;
        StringBuilder ans = new StringBuilder();
        int n = s.length();
        for(int i=0; i<n; ++i){
            char c = s.charAt(i);
            if(open==0 && c==')')
                continue;
            if(c == '(')
                open++;
            if(c==')')
                open--;
            ans.append(c);
        }
        String ans1 = ans.toString();
        n = ans1.length();

        StringBuilder ans2 = new StringBuilder();
        int close = 0;
        for(int i=n-1; i>=0; --i){
            char c = ans.charAt(i);
            if(close==0 && c=='(')
                continue;
            if(c == ')')
                close++;
            if(c=='(')
                close--;
            ans2.append(c);
        }

        return ans2.reverse().toString();
    }
}